go/cmd/dolt/commands/blame.go

```diff
@@ -0,0 +1,450 @@
+	}
+	// if the row is in the new table but not the old one, it must be new; return true
+	if newRow != nil && oldRow == nil {
+		return true, nil
+	}
+
+	oldSchema, err := oldTable.GetSchema(ctx)
+	if err != nil {
+		return false, fmt.Errorf("error getting schema from %s in old commit %s: %v", tableName, oldHash, err)
+	}
+
+	return !row.AreEqual(*oldRow, *newRow, oldSchema), nil
+}
+
+func blameGraphFromRows(ctx context.Context, rows types.Map) (*blameGraph, error) {
+	graph := make(blameGraph)
+	err := rows.IterAll(ctx, func(key, _ types.Value) error {
+		hash, err := key.Hash(types.Format_7_18)
+		if err != nil {
+			return err
+		}
+		graph[hash] = blameInfo{Key: key}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &graph, nil
+}
+
+// UnblamedNodeKeys returns a slice of primary keys which do not have blame information
+func (bg *blameGraph) UnblamedNodeKeys() []types.Value {
+	keys := []types.Value{}
+	for _, v := range *bg {
+		if v.CommitHash == "" {
+			keys = append(keys, v.Key)
+		}
+	}
+	return keys
+}
+
+// AllNodesBlamed returns true if there are no unblamed nodes in the blame graph
+func (bg *blameGraph) AllNodesBlamed() bool {
+	return len(bg.UnblamedNodeKeys()) == 0
+}
+
+// BlameNode updates the blame graph to contain blame information from the given commit
+// for the row identified by the given primary key
+func (bg *blameGraph) BlameNode(rowPK types.Value, c *doltdb.Commit) error {
+	commitHash, err := c.HashOf()
+	if err != nil {
+		return fmt.Errorf("error getting commit hash: %v", err)
+	}
+
+	meta, err := c.GetCommitMeta()
+	if err != nil {
+		return fmt.Errorf("error getting metadata for commit %s: %v", commitHash.String(), err)
+	}
+
+	pkHash, err := rowPK.Hash(types.Format_7_18)
+	if err != nil {
+		return fmt.Errorf("error getting PK hash for commit %s: %v", commitHash.String(), err)
+	}
+
+	(*bg)[pkHash] = blameInfo{
+		Key:         rowPK,
+		CommitHash:  commitHash.String(),
+		Author:      meta.Name,
+		Description: meta.Description,
+		Timestamp:   meta.Timestamp,
+	}
+
+	return nil
+}
+
+func getPKStrs(ctx context.Context, pk types.Value) (strs []string) {
+	i := 0
+	pk.WalkValues(ctx, func(val types.Value) error {
+		// even-indexed values are index numbers. they aren't useful, don't print them.
+		if i%2 == 1 {
+			strs = append(strs, fmt.Sprintf("%v", val))
+		}
+		i++
+		return nil
+	})
+
+	return strs
+}
+
+func truncateString(str string, maxLength int) string {
+	if maxLength < 0 || len(str) <= maxLength {
+		return str
+	}
+
+	if maxLength == 0 || str == "" {
+		return ""
+	}
+
+	if maxLength == 1 {
+		return "…"
+	}
+
+	return str[0:maxLength-1] + "…"
+}
+
+var dataColNames = []string{"Commit Msg", "Author", "Time", "Commit"}
+
+// String returns the string representation of this blame graph
+func (bg *blameGraph) String(ctx context.Context, pkColNames []string) string {
+	// here we have two []string and need one []interface{} (aka table.Row)
```

Comment:
Seems right to me. A slight variation that comes to mind:

```go
header := []interface{}{}
for _, cellText := range append(pkColNames, dataColNames...) {
	header = append(header, cellText)
}
```

Reply:
Rad, thanks.